pss/pss.go

```diff
@@ -219,21 +287,29 @@ func (p *Pss) Start(srv *p2p.Server) error {
 			}
 		}
 	}()
-	go func() {
-		for {
-			select {
-			case msg := <-p.outbox:
-				metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(len(p.outbox)))
 
-				err := p.forward(msg.msg)
-				if err != nil {
-					log.Error(err.Error())
-					metrics.GetOrRegisterCounter("pss.forward.err", nil).Inc(1)
+	// In any case, if the message
+	// forwarding fails, the node should try to forward it to the next best peer, until the message is
+	// successfully forwarded to at least one peer.
+	go func() {
+		for slot := range p.outbox.process {
+			go func(slot int) {
+				msg := p.outbox.msg(slot)
+				sent := p.forwardFunc(msg.msg)
+				if sent {
+					// free the outbox slot
+					p.outbox.free(slot)
+					metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(p.outbox.len()))
+				} else {
+					// if we failed to send to anyone, re-insert message in the send-queue
+					log.Debug("unable to forward to any peers", "slot", slot)
+					// reenqueue the message for processing
+					p.outbox.reenqueue(slot)
+					log.Debug("Message reenqued", "slot", slot)
 				}
 				metrics.GetOrRegisterResettingTimer("pss.handle.outbox", nil).UpdateSince(msg.startedAt)
-			case <-p.quitC:
-				return
-			}
+
```

Comment:
should this metrics line not be at the top of the for loop?

Reply:
I don't know much about these metrics, but it was at the bottom of the method in previous version. But maybe you are right and should be updated as soon as a slot is received through the `process` channel.

Reply:
Moved metric to the top of the method in 1b91b66